### PR TITLE
Introduce new precision layout in PL

### LINKED ISSUE #9956: Is `precision="mixed"` redundant?
## Proposed refactoring or deprecation

Does `precision="mixed"` act differently to `precision=16` in any way?

I understand that "mixed" is more correct as 16-bit precision can still run some computations in 32-bit.

### Motivation

In https://github.com/PyTorchLightning/pytorch-lightning/pull/9763 I noticed we did not even have a `PrecisionType` for `"mixed"`.

There's a single test in the codebase passing the "mixed" value:

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/plugins/test_deepspeed_plugin.py#L153

And no mention at all of this value in the docs.

### Pitch

Have one value to set this, whether it is `16` or `"mixed"`. Most likely `16` since its the one widely used.

Otherwise, add tests for passing `"mixed"`

### Additional context

```python
$ grep -iIrn '"mixed"' pytorch_lightning
pytorch_lightning/plugins/training_type/sharded.py:62:                    is_fp16 = precision in ("mixed", 16)
pytorch_lightning/plugins/training_type/fully_sharded.py:135:            mixed_precision=precision == "mixed",
pytorch_lightning/plugins/training_type/ipu.py:42:        if self.precision in ("mixed", 16):
pytorch_lightning/plugins/training_type/deepspeed.py:405:            dtype = torch.float16 if self.precision in (16, "mixed") else torch.float32
pytorch_lightning/plugins/training_type/deepspeed.py:473:            dtype = torch.float16 if self.precision in (16, "mixed") else torch.float32
pytorch_lightning/plugins/training_type/deepspeed.py:602:        if precision in (16, "mixed"):
pytorch_lightning/plugins/precision/mixed.py:26:    precision: Union[str, int] = "mixed"
pytorch_lightning/plugins/precision/fully_sharded_native_amp.py:26:    precision = "mixed"
```

```python
$ grep -iIrn '"mixed"' tests
tests/plugins/test_deepspeed_plugin.py:153:@pytest.mark.parametrize("precision", [16, "mixed"])
```

```python
$ grep -Irn 'mixed' docs | grep 'precision='
# no mention in the docs!
```
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

cc @justusschock @awaelchli @akihironitta @rohitgr7 @tchaton @borda @carmocca

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Adds the new precision inputs from #16767 to PL.
From now on, the plugins and strategies internally only care about the new values (`'64-true'`, `'32-true'`, `'16-mixed'`, `'bf16-mixed'`) while the connector translates all other supported values (`64`, `'64'`, `32`, `'32'`, `16`, `'16'`, `'bf16'`) to these values and raises a warning for the mixed precision ones accordingly.

Fixes #9956 

There will be a follow-up PR to re-enable testing the external colossal-ai strategy once these changes are reflected and released within this strategy.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @carmocca @justusschock @awaelchli

### CODE REVIEW

File: tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py [hunk 1]
```diff
@@ -837,15 +839,15 @@ def get_defaults(cls):
...
-    with mock.patch(
-        "lightning.pytorch.trainer.connectors.accelerator_connector._LIGHTNING_COLOSSALAI_AVAILABLE", False
-    ), pytest.raises(ModuleNotFoundError):
-        Trainer(strategy="colossalai")
-
-    from lightning_colossalai import ColossalAIStrategy
-
-    trainer = Trainer(strategy="colossalai", precision=16)
-    assert isinstance(trainer.strategy, ColossalAIStrategy)
+# @RunIf(min_cuda_gpus=1)  # trigger this test on our GPU pipeline, because we don't install the package on the CPU suite
```
Comment (justusschock) [inline tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py:841]: This is commented out here, since the validation for the precision argument is hardcoded in https://github.com/Lightning-AI/lightning-colossalai . I tried to first change it there, but these changes won't become effective until the strategy was released to pypi again.

After this was merged, I will post a follow-up PR in https://github.com/Lightning-AI/lightning-colossalai which will then be released and then have a follow-up PR to again add this test.

Discussed offline with @awaelchli

File: tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py [hunk 2]
```diff
@@ -839,6 +841,7 @@ def get_defaults(cls):
 
 @RunIf(min_cuda_gpus=1)  # trigger this test on our GPU pipeline, because we don't install the package on the CPU suite
 @pytest.mark.skipif(not package_available("lightning_colossalai"), reason="Requires Colossal AI Strategy")
+@pytest.mark.skip
```
Comment (justusschock) [inline tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py:844]: This is skipped here, since the validation for the precision argument is hardcoded in https://github.com/Lightning-AI/lightning-colossalai . I tried to first change it there, but these changes won't become effective until the strategy was released to pypi again.

After this was merged, I will post a follow-up PR in https://github.com/Lightning-AI/lightning-colossalai which will then be released and then have a follow-up PR to again add this test.

Discussed offline with @awaelchli

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
Awesome!
